edgelet/edgelet-core/src/module.rs

```diff
@@ -266,21 +269,102 @@ pub trait ModuleRegistry {
 }
 
-#[derive(Debug, Default, Deserialize, PartialEq, Serialize)]
+#[skip_serializing_none]
+#[derive(Debug, Default, PartialEq, Serialize)]
+#[serde(rename_all = "camelCase")]
 pub struct SystemInfo {
-    /// OS Type of the Host. Example of value expected: \"linux\" and \"windows\".
-    #[serde(rename = "osType")]
-    pub os_type: String,
-    /// Hardware architecture of the host. Example of value expected: arm32, x86, amd64
+    pub kernel: String,
+    pub kernel_release: String,
+    pub kernel_version: String,
+
+    pub operating_system: Option<String>,
+    pub operating_system_version: Option<String>,
+    pub operating_system_variant: Option<String>,
+    pub operating_system_build: Option<String>,
+
     pub architecture: String,
-    /// iotedge version string
+    pub cpus: usize,
+    pub virtualized: String,
+
+    pub product_name: Option<String>,
+    pub system_vendor: Option<String>,
+
     pub version: String,
     pub provisioning: ProvisioningInfo,
-    pub server_version: String,
-    pub kernel_version: String,
-    pub operating_system: String,
-    pub cpus: i32,
-    pub virtualized: String,
+
+    #[serde(default, flatten)]
+    pub additional_properties: BTreeMap<String, String>,
+}
+
+impl SystemInfo {
+    pub fn from_system() -> Result<Self, Error> {
+        let kernel = nix::sys::utsname::uname();
+        let dmi = DmiInfo::default();
+        let os = OsInfo::default();
+
+        let res = Self {
+            kernel: kernel.sysname().to_owned(),
+            kernel_release: kernel.release().to_owned(),
+            kernel_version: kernel.version().to_owned(),
+
+            operating_system: os.id,
+            operating_system_version: os.version_id,
+            operating_system_variant: os.variant_id,
+            operating_system_build: os.build_id,
+
+            architecture: os.arch.to_owned(),
+            cpus: num_cpus::get(),
+            virtualized: match crate::virtualization::is_virtualized_env() {
+                Ok(Some(true)) => "yes",
+                Ok(Some(false)) => "no",
+                _ => "unknown",
+            }
+            .to_owned(),
+
+            product_name: dmi.product,
+            system_vendor: dmi.vendor,
+
+            version: crate::version_with_source_version(),
+            provisioning: ProvisioningInfo {
+                r#type: "ProvisioningType".into(),
+                dynamic_reprovisioning: false,
+                always_reprovision_on_startup: false,
+            },
+
+            additional_properties: BTreeMap::new(),
+        };
+
+        Ok(res)
+    }
+
+    pub fn merge_additional(&mut self, mut additional_info: BTreeMap<String, String>) -> &Self {
+        macro_rules! remove_assign {
+            ($src:literal, $dest:ident) => {
+                if let Some((_, x)) = additional_info.remove_entry($src) {
+                    self.$dest = x.into();
+                }
+            };
+        }
+
+        remove_assign!("kernel_name", kernel);
```

Comment:
I am not entirely following this part. Is that a backdoor to allow user to override the defaults?
If that's the case, you should document what device info are present as a default and mention that you can override them in the config.